File: pract41/main.cpp
#include <iostream>
#include <stack>
#include <vector>
#include <sstream>
#include <cctype>
#include <map>
#include <regex>
using namespace std;

// Operator precedence
map<char, int> precedence = {
    {'^', 3},
    {'*', 2}, {'/', 2},
    {'+', 1}, {'-', 1}
};

// Check if a character is an operator
bool isOperator(char c) {
    return precedence.count(c);
}

// Split string into tokens (variables, operators, parentheses)
vector<string> tokenize(const string& expr) {
    vector<string> tokens;
    regex token_re(R"([a-zA-Z_]\w*|\d+|[+*/^()-])");
    auto words_begin = sregex_iterator(expr.begin(), expr.end(), token_re);
    auto words_end = sregex_iterator();
    for (auto it = words_begin; it != words_end; ++it)
        tokens.push_back(it->str());
    return tokens;
}

// Convert infix to postfix (Shunting Yard Algorithm)
vector<string> infixToPostfix(const string& expr) {
    stack<string> opStack;
    vector<string> postfix;
    vector<string> tokens = tokenize(expr);

    for (string token : tokens) {
        if (isalnum(token[0])) {
            postfix.push_back(token);
        }
        else if (token == "(") {
            opStack.push(token);
        }
        else if (token == ")") {
            while (!opStack.empty() && opStack.top() != "(") {
                postfix.push_back(opStack.top());
                opStack.pop();
            }
            if (!opStack.empty()) opStack.pop(); // pop "("
        }
        else if (isOperator(token[0])) {
            while (!opStack.empty() && opStack.top() != "(" &&
                   precedence[token[0]] <= precedence[opStack.top()[0]]) {
                postfix.push_back(opStack.top());
                opStack.pop();
            }
            opStack.push(token);
        }
    }

    while (!opStack.empty()) {
        postfix.push_back(opStack.top());
        opStack.pop();
    }

    return postfix;
}

// Generate Three Address Code
void generateTAC(const vector<string>& postfix, const string& lhs) {
    stack<string> evalStack;
    int tempCount = 1;
    vector<string> code;

    for (string token : postfix) {
        if (isalnum(token[0])) {
            evalStack.push(token);
        }
        else if (isOperator(token[0])) {
            string op2 = evalStack.top(); evalStack.pop();
            string op1 = evalStack.top(); evalStack.pop();
            string temp = "t" + to_string(tempCount++);
            cout << temp << " = " << op1 << " " << token << " " << op2 << endl;
            evalStack.push(temp);
        }
    }

    cout << lhs << " = " << evalStack.top() << endl;
}

// Main function
int main() {
    string input;
    cout << "Enter an expression (e.g., a = b*b + c*c): ";
    getline(cin, input);

    size_t eq = input.find('=');
    if (eq == string::npos) {
        cout << "Invalid expression format. Use: variable = expression" << endl;
        return 1;
    }

    string lhs = input.substr(0, eq);
    string rhs = input.substr(eq + 1);

    lhs.erase(remove(lhs.begin(), lhs.end(), ' '), lhs.end());
    vector<string> postfix = infixToPostfix(rhs);
    
    cout << "\nThree Address Code:\n";
    generateTAC(postfix, lhs);

    return 0;
}

File: pract41/main.py
import re

# Define operator precedence
precedence = {
    '^': 3,
    '*': 2, '/': 2,
    '+': 1, '-': 1
}

# Check if character is an operator
def is_operator(c):
    return c in precedence

# Convert infix to postfix using shunting yard algorithm
def infix_to_postfix(expression):
    stack = []
    output = []
    tokens = re.findall(r"[a-zA-Z_]\w*|\d+|[+*/^()-]", expression)

    for token in tokens:
        if token.isalnum():
            output.append(token)
        elif token == '(':
            stack.append(token)
        elif token == ')':
            while stack and stack[-1] != '(':
                output.append(stack.pop())
            stack.pop()  # pop '('
        elif is_operator(token):
            while (stack and stack[-1] != '(' and
                   precedence.get(token, 0) <= precedence.get(stack[-1], 0)):
                output.append(stack.pop())
            stack.append(token)
    while stack:
        output.append(stack.pop())

    return output

# Generate Three Address Code from postfix
def generate_TAC(postfix, result_var):
    stack = []
    temp_count = 1
    code = []

    for token in postfix:
        if token.isalnum():
            stack.append(token)
        elif is_operator(token):
            op2 = stack.pop()
            op1 = stack.pop()
            temp_var = f't{temp_count}'
            temp_count += 1
            code.append(f"{temp_var} = {op1} {token} {op2}")
            stack.append(temp_var)

    final_result = stack.pop()
    code.append(f"{result_var} = {final_result}")
    return code

# Main function
def main():
    expr = input("Enter an expression (e.g., a = b*b + c*c): ").strip()

    if '=' not in expr:
        print("Invalid expression format. Use: variable = expression")
        return

    lhs, rhs = map(str.strip, expr.split('='))
    postfix = infix_to_postfix(rhs)
    tac = generate_TAC(postfix, lhs)

    print("\nThree Address Code:")
    for line in tac:
        print(line)

if __name__ == "__main__":
    main()
